### ISSUE TITLE
after `docker swarm init` daemon does not listen on public IP until after `service docker restart`

### ISSUE THREAD

**Issue by ijc25** [docker/docker#23672]
**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.0-rc1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   1f136c1-unsupported
 Built:        Wed Jun 15 05:00:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   1f136c1-unsupported
 Built:        Wed Jun 15 05:00:08 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
$ docker info
Containers: 8
 Running: 0
 Paused: 0
 Stopped: 8
Images: 1037
Server Version: 1.12.0-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 876
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay host null
Swarm: active
 NodeID: b4rtwxf74v6lu02br9y9u5w7s
 IsManager: Yes
 Managers: 1
 Nodes: 1
 CACertHash: sha256:a83be2b60df19147ae729c1f3d5684879ad03fae4d34b2f6dcfaf48884b1cbc8
Runtimes: default
Default Runtime: default
Kernel Version: 4.6.0-1-amd64
Operating System: Debian GNU/Linux stretch/sid
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.55 GiB
Name: bokrug
ID: AU33:BO7D:7VGM:MOLB:RSDF:IBRV:GCWT:THHM:OEVM:TX3C:BNLM:WHAR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 37
 Goroutines: 118
 System Time: 2016-06-17T15:01:13.99192859+01:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Debian Linux (x86), using packages from:
```
deb https://apt.dockerproject.org/repo debian-stretch testing
```

**Steps to reproduce the issue:**
1. start with docker without swarm initialised
2. `sudo netstat -nap | grep ^tcp.*dockerd` to confirm it listens on no TCP ports
3. `docker swarm init`
4. `sudo netstat -nap | grep ^tcp.*dockerd` and observe dockerd listening to some tcp ports, but not 2377 on external IP (172.16.10.41 for this machine):
```
tcp        0      0 172.16.10.41:7946       0.0.0.0:*               LISTEN      17316/dockerd       
tcp        0      0 127.0.0.1:39872         127.0.0.1:2377          ESTABLISHED 17316/dockerd       
tcp6       0      0 :::2377                 :::*                    LISTEN      17316/dockerd       
tcp6       0      0 127.0.0.1:2377          127.0.0.1:39872         ESTABLISHED 17316/dockerd 
```
5. `sudo service docker restart`
6. `sudo netstat -nap | grep ^tcp.*dockerd` and observe dockerd listening to additional ports, including 2377 on external IP:
```
tcp        0      0 172.16.10.41:7946       0.0.0.0:*               LISTEN      17824/dockerd       
tcp        0      0 172.16.10.41:33862      172.16.10.41:2377       ESTABLISHED 17824/dockerd       
tcp        0      0 172.16.10.41:33868      172.16.10.41:2377       ESTABLISHED 17824/dockerd       
tcp6       0      0 :::2377                 :::*                    LISTEN      17824/dockerd       
tcp6       0      0 172.16.10.41:2377       172.16.10.41:33862      ESTABLISHED 17824/dockerd       
tcp6       0      0 172.16.10.41:2377       172.16.10.41:33868      ESTABLISHED 17824/dockerd       
```

**Describe the results you received:**

Inline in the repro steps above.

**Describe the results you expected:**

I expected the set of externally listening ports to be the same before and after the `service docker restart`, i.e. step *#*4 and *#*6 the same.

AIUI I think I should have expected the ports to appear after the `docker swarm init`, i.e. *#*4 should have been listening on more ports rather than *#*6 listening on fewer.

It's possible that instead I should have expected no ports at all in *#*4 (i.e. the same as *#*2) and for the full set to show up in *#*6, but that's no my understanding of how it is expected to work.

In any case listening to some but not all ports in step *#*4 is unexpected.

**Additional information you deem important (e.g. issue happens only occasionally):**

I have also tried this in a docker 4 mac environment with https://github.com/justincormack/docker/commits/moby-1.12.0-rc2 which is the set of commits closer to what will be rc2, the results were the same. `docker version` in that context:
````
# docker version
Client:
 Version:      1.12.0-rc1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   244b115
 Built:        Fri Jun 17 13:24:58 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   244b115
 Built:        Fri Jun 17 13:24:58 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Comment by ijc25** [comment]
Another experiment: `docker swarm init --listen-addr 172.16.10.41:2733` seems to do as I would expect and starts listening right away:
```
$ sudo netstat -nap | grep ^tcp.*dockerd
tcp        0      0 172.16.10.41:7946       0.0.0.0:*               LISTEN      17824/dockerd       
tcp        0      0 172.16.10.41:2733       0.0.0.0:*               LISTEN      17824/dockerd       
tcp        0      0 172.16.10.41:2733       172.16.10.41:32904      ESTABLISHED 17824/dockerd       
tcp        0      0 172.16.10.41:32904      172.16.10.41:2733       ESTABLISHED 17824/dockerd      
```

`docker swarm init --listen-addr 0.0.0.0:2733` behaves wrongly as described above (no surprise given that 0.0.0.0:2733 is the default).

**Comment by mgoelzer** [comment]
ping @tonistiigi

**Comment by cpuguy83** [comment]
I see you had this issue with rc1. Did you have the same issue with rc2?

**Comment by tonistiigi** [comment]
Not sure if I'm missing something but `:::2377` should mean that it is listening on all interfaces. `172.16.10.41:2377       172.16.10.41:33862` in your second example is not listening but actual connection, these only appear if you try to connect to that address. I think the difference after restart is that the worker connects back to manager using manager's advertised address(public ip) instead of localhost that it used the first time.

**Comment by ijc25** [comment]
@cpuguy83 I did try with "almost rc2" at the time (since rc2 wasn't out then) via docker 4 mac and it did suffer from the same issue.

@tonistiigi perhaps I have become confused by forgetting that `:::2377` can in some circumstances also include all IPv4 interfaces and/or by misreading the `LISTEN` vs `ESTABLISHED` lines. The reason I was looking at this in the first place was that I was trying to get my docker 4 mac to join a swarm with my Linux laptop, so it's also possible I became further confused by some misbehaviour or other on the other end there and latched onto this as a red-herring.

I'll give all this another go with the latest stuff and either figure out where I went wrong or be able to provide a more coherent idea of where things went wrong.
